pkg/gcpwatch/tracker.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcpwatch
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/apis/iam/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+type DependencyTracker struct {
+	fetcher Fetcher
+
+	gcpResources map[gcpResouceKey]*dependenciesByResource
+
+	controllersMutex sync.Mutex
+	controllers      map[string]*ControllerRegistration
+}
+
+type ControllerRegistration struct {
+	tracker       *DependencyTracker
+	controllerKey string
+	queue         chan event.GenericEvent
+}
+
+func (t *DependencyTracker) RegisterController(key string, queue chan event.GenericEvent) *ControllerRegistration {
+	t.controllersMutex.Lock()
+	defer t.controllersMutex.Unlock()
+
+	if _, exists := t.controllers[key]; exists {
+		klog.Fatalf("multiple controllers with key %q", key)
+	}
+
+	registration := &ControllerRegistration{
+		controllerKey: key,
+		queue:         queue,
+		tracker:       t,
+	}
+
+	t.controllers[key] = registration
+
+	return registration
+}
+
+type gcpResouceKey struct {
+	Kind     string
+	External string
+}
+
+type dependenciesByResource struct {
+	etag string
```

Comment:
Not going to block this PR but I think this is wrong. We should have a more abstract piece here. Etag does not work on any resource which we expect to be regularly updated (Eg it has something like a heartbeat). We should have something more generic and etag as the default implementation. However some resources will probably require us to hash a mask of the resource values.